src/suction_grasp_generator.cpp

```diff
@@ -0,0 +1,413 @@
+                                     const Eigen::Vector3d& object_size, double object_width,
+                                     std::vector<GraspCandidatePtr>& grasp_candidates)
+{
+  // Transform the grasp pose eef mount to the tcp grasp pose
+  Eigen::Isometry3d grasp_pose_tcp = grasp_pose_eef_mount * grasp_data->tcp_to_eef_mount_.inverse();
+
+  // The new grasp
+  moveit_msgs::Grasp new_grasp;
+
+  // Approach and retreat - aligned with eef to grasp transform
+  // set pregrasp
+  moveit_msgs::GripperTranslation pre_grasp_approach;
+  new_grasp.pre_grasp_approach.direction.header.stamp = ros::Time::now();
+  new_grasp.pre_grasp_approach.desired_distance = grasp_data->grasp_max_depth_ + grasp_data->approach_distance_desired_;
+  new_grasp.pre_grasp_approach.min_distance = 0;  // NOT IMPLEMENTED
+  new_grasp.pre_grasp_approach.direction.header.frame_id = grasp_data->parent_link_->getName();
+
+  Eigen::Vector3d grasp_approach_vector = -1 * grasp_data->tcp_to_eef_mount_.translation();
+  grasp_approach_vector = grasp_approach_vector / grasp_approach_vector.norm();
+
+  new_grasp.pre_grasp_approach.direction.vector.x = grasp_approach_vector.x();
+  new_grasp.pre_grasp_approach.direction.vector.y = grasp_approach_vector.y();
+  new_grasp.pre_grasp_approach.direction.vector.z = grasp_approach_vector.z();
+
+  // set postgrasp
+  moveit_msgs::GripperTranslation post_grasp_retreat;
+  new_grasp.post_grasp_retreat.direction.header.stamp = ros::Time::now();
+  new_grasp.post_grasp_retreat.desired_distance = grasp_data->grasp_max_depth_ + grasp_data->retreat_distance_desired_;
+  new_grasp.post_grasp_retreat.min_distance = 0;  // NOT IMPLEMENTED
+  new_grasp.post_grasp_retreat.direction.header.frame_id = grasp_data->parent_link_->getName();
+  new_grasp.post_grasp_retreat.direction.vector.x = -1 * grasp_approach_vector.x();
+  new_grasp.post_grasp_retreat.direction.vector.y = -1 * grasp_approach_vector.y();
+  new_grasp.post_grasp_retreat.direction.vector.z = -1 * grasp_approach_vector.z();
+
+  // set grasp pose
+  geometry_msgs::PoseStamped grasp_pose_msg;
+  grasp_pose_msg.header.stamp = ros::Time::now();
+  grasp_pose_msg.header.frame_id = grasp_data->base_link_;
+
+  // name the grasp
+  static std::size_t grasp_id = 0;
+  new_grasp.id = "Grasp" + boost::lexical_cast<std::string>(grasp_id);
+  grasp_id++;
+
+  tf::poseEigenToMsg(grasp_pose_eef_mount, grasp_pose_msg.pose);
+  new_grasp.grasp_pose = grasp_pose_msg;
+
+  // set grasp postures e.g. hand closed
+  new_grasp.grasp_posture = grasp_data->grasp_posture_;
+
+  std::vector<double> suction_voxel_overlap;
+  new_grasp.grasp_quality =
+      scoreSuctionGrasp(grasp_pose_tcp, grasp_data, object_pose, object_size, suction_voxel_overlap);
+
+  auto suction_grasp_candidate = std::make_shared<SuctionGraspCandidate>(new_grasp, grasp_data, object_pose);
+  suction_grasp_candidate->suction_voxel_overlap_ = suction_voxel_overlap;
+  grasp_candidates.emplace_back(std::move(suction_grasp_candidate));
```

Comment:
This seems like overkill for something with almost no performance benefit over
`grasp_candidates.push_back(suction_grasp_candidate)`

Reply:
if you want to use emplace_back without move you could do this:
```c++
  grsp_candidates.emplace_baack(std::make_shared<SuctionGraspCandidate>(new_grasp, grasp_data, object_pose));
  grasp_candidates.back()->suction_voxel_overlap_ = suction_voxel_overlap;
```

Reply:
There is no performance benefit. It is exactly the same, so might as well use push_back().

Reply:
Is there any downside to this?